dens-validation/src/MyLib.hs

```diff
@@ -1,41 +1,276 @@
+    pguardM "ElementID token minted" $
+        (pvalueOf # fields.mint # setElemIdCS # pcon (PTokenName $ pconstant ""))
+            #== 1
+
+    pcon PUnit
+
+mkSetElemMintingPolicy ::
+    Term
+        s
+        ( PAddress -- Address of the (unspendable) protocol validator
+            :--> PCurrencySymbol -- Protocol NFT (assume empty token name for now)
+            :--> SetInsert
+            :--> PScriptContext
+            :--> PUnit
+        )
+mkSetElemMintingPolicy = phoistAcyclic $ plam $ \protocolAddr protocolSymb setInsert cxt -> P.do
+    SetInsert'Insert keyToInsert <- pmatch setInsert
+
+    txInfo <- plet $ pfield @"txInfo" # cxt
+
+    fields <- pletFields @'["inputs", "referenceInputs", "outputs", "mint"] txInfo
+
+    -- 0) REFERENCE INPUTS: Get the protocol datum (reference input, paid to protocol validator)
+
+    Protocol elemIdMP setElemMP setValidator _ <-
+        pmatch $
+            extractDatum @Protocol
+                # protocolAddr
+                # protocolSymb
+                #$ pmap
+                # resolved
+                # fields.referenceInputs
+
+    -- 1) INPUTS: Check for k < densKey < nxt input that pays to the set validator & holds a
+
+    -- not strictly necessary but can't hurt
+    pguardM "Only one input" (plength # pfromData fields.inputs #== 1)
+    setValidatorAddress <- plet $ pscriptHashAddress # setValidator
+    setElemCS <- plet $ scriptHashToCS # pfromData setElemMP -- tx in inputs should have SetElem NFT (right?)
+    setDatum <- plet $ extractDatum @SetDatum # setValidatorAddress # setElemCS #$ pmap # resolved # fields.inputs -- ignoring OwnerApproval for now
+    SetDatum l r _ <- pmatch setDatum
+    pguardM "Validate set insert" $ validateSetInsert # setDatum # pfromData keyToInsert
+
+    -- 2) OUTPUTS: Check for a. SD(k, densKey) b. SD(densKey,nxt)
+    outputs <- plet $ pfromData fields.outputs
+    pguardM "Only two outputs" (plength # outputs #== 2)
+    l' <- plet $ pfromData l
+    r' <- plet $ pfromData r
+    k <- plet $ pfromData keyToInsert
+    checkOutput <- plet $ plam $ \p -> extractDatumSuchThat @SetDatum # p # setValidatorAddress # setElemCS # outputs
+    SetDatum{} <- pmatch $ checkOutput # (hasLR # l' # k)
+    SetDatum{} <- pmatch $ checkOutput # (hasLR # k # r')
+
+    -- 3) MINTS: Check for the presence of a SetElemID NFT & ElementID NFT
+
+    mint <- plet $ pnormalize # pfromData fields.mint
+
+    checkMintsOne <- plet $ plam $ \currSym tokName -> (pvalueOf # mint # currSym # tokName) #== 1
+
+    pguardM "Mints one SetElemID token" $ checkMintsOne # setElemCS # emptyTN
+
+    elemIdCS <- plet $ scriptHashToCS # pfromData elemIdMP
+    kTokName <- plet $ pcon $ PTokenName (pblake2b_256 #$ pserialiseData # pdataImpl k)
+    pguardM "Mints one ElementID NFT with a token name == blake2b_256(densKey)" $ checkMintsOne # elemIdCS # kTokName
+    -- TODO: We probably want to double check that nothing *else* gets minted (but that's annoying to implement)
```

Comment:
Yes, we need this check actually. Please implement this. 

Since otherwise, anyone can mint their own `ElementID` NFTs and "own" anything.